### PR TITLE
Expose dynamic_templates parameter in Ingest

### PR DESCRIPTION
**Pull request by dnhatn**
This change exposes the newly introduced parameter `dynamic_templates` in ingest. This parameter can be set by a set processor or a script processor  (see the YAML test).

Relates #69948

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 1]
```diff
@@ -652,6 +652,11 @@ private void innerExecute(int slot, IndexRequest indexRequest, Pipeline pipeline
                     indexRequest.setIfPrimaryTerm(((Number) metadataMap.get(IngestDocument.Metadata.IF_PRIMARY_TERM)).longValue());
                 }
                 indexRequest.source(ingestDocument.getSourceAndMetadata(), indexRequest.getContentType());
+                if (metadataMap.get(IngestDocument.Metadata.DYNAMIC_TEMPLATES) != null) {
+                    Map<String, String> mergedDynamicTemplates = new HashMap<>(indexRequest.getDynamicTemplates());
+                    mergedDynamicTemplates.putAll((Map<String, String>) metadataMap.get(IngestDocument.Metadata.DYNAMIC_TEMPLATES));
+                    indexRequest.setDynamicTemplates(mergedDynamicTemplates);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:658]: can we add docs?  As a user, without looking at the code, I would wonder if I end up overwriting what has been set by the bulk request. Also, whether I have access to the templates that are already set or I can only provide new ones.

**Review by javanna — COMMENTED** [review]
I left a comment on behaviour and docs. The change makes sense to me.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:658]: ++, I added a doc in https://github.com/elastic/elasticsearch/pull/71716/commits/41f08b78dbfeb95aea616cc377556a34ba093055.

File: docs/reference/ingest.asciidoc [hunk 2]
```diff
@@ -539,6 +540,29 @@ PUT _ingest/pipeline/my-pipeline
...
+          "address": "geo_point"       
+        }
+      }
+    }
+  ]
+}
+----
+
+The set processor above tells ES to use the dynamic template named `geo_point`
+for the field `address` (if this field does not exist in the index). Note that
```
Comment (javanna) [inline docs/reference/ingest.asciidoc:562]: shall we say if the field is not mapped yet instead?

File: docs/reference/ingest.asciidoc [hunk 3]
```diff
@@ -539,6 +540,29 @@ PUT _ingest/pipeline/my-pipeline
...
+      }
+    }
+  ]
+}
+----
+
+The set processor above tells ES to use the dynamic template named `geo_point`
+for the field `address` (if this field does not exist in the index). Note that 
+this processor overrides only the `_dynamic_templates` of the field `address`
+but leaves the settings of other fields intact.
```
Comment (javanna) [inline docs/reference/ingest.asciidoc:564]: I think that we should try and clarify this sentence: "this processor overrides the dynamic template for the field `address` if already defined in the bulk request, but has no effect on other dynamic templates defined in the bulk request."

**Review by javanna — APPROVED** [review]
left a small comment on docs, LGTM otherwise

File: docs/reference/ingest.asciidoc (hunk 2 above)
Comment (dnhatn) [reply docs/reference/ingest.asciidoc]: adjusted in https://github.com/elastic/elasticsearch/pull/71716/commits/6245e1e8e276413ce968daaa3a2f25873c03c92b

File: docs/reference/ingest.asciidoc (hunk 3 above)
Comment (dnhatn) [reply docs/reference/ingest.asciidoc]: fixed in https://github.com/elastic/elasticsearch/pull/71716/commits/6245e1e8e276413ce968daaa3a2f25873c03c92b

**Comment by dnhatn** [comment]
@javanna Thanks for reviewing.

**Pull request merged**